File: android/app/src/main/java/com/dailysns/viewmodels/PostCreationViewModel.kt
package com.dailysns.viewmodels

import androidx.lifecycle.ViewModel
import com.dailysns.models.Post
import com.dailysns.models.PostItem
import com.dailysns.models.PostItemType
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * PostCreationViewModel.kt
 * DailySNSWithYourSS Android版
 * 
 * 投稿作成のビジネスロジック
 */

class PostCreationViewModel : ViewModel() {
    
    private val _postItems = MutableStateFlow<List<PostItem>>(emptyList())
    val postItems: StateFlow<List<PostItem>> = _postItems.asStateFlow()
    
    private val _isCreatingPost = MutableStateFlow(false)
    val isCreatingPost: StateFlow<Boolean> = _isCreatingPost.asStateFlow()
    
    private val _currentItemType = MutableStateFlow(PostItemType.SCREENSHOT)
    val currentItemType: StateFlow<PostItemType> = _currentItemType.asStateFlow()
    
    /**
     * スクリーンショットを追加
     */
    fun addScreenshot(imageURL: String) {
        val item = PostItem(type = PostItemType.SCREENSHOT, content = imageURL)
        _postItems.value = _postItems.value + item
        _currentItemType.value = PostItemType.COMMENT
    }
    
    /**
     * コメントを追加
     */
    fun addComment(text: String) {
        if (text.isBlank()) return
        val item = PostItem(type = PostItemType.COMMENT, content = text)
        _postItems.value = _postItems.value + item
        _currentItemType.value = PostItemType.SCREENSHOT
    }
    
    /**
     * アイテムを削除（インデックス指定）
     */
    fun removeItemAt(index: Int) {
        if (index in _postItems.value.indices) {
            _postItems.value = _postItems.value.toMutableList().apply {
                removeAt(index)
            }
        }
    }
    
    /**
     * アイテムを削除（ID指定）
     */
    fun removeItem(id: String) {
        _postItems.value = _postItems.value.filter { it.id != id }
    }
    
    /**
     * アクティブなアイテムを取得
     */
    private fun getActiveItems(): List<PostItem> {
        return _postItems.value.filter { !it.isDeleted }
    }
    
    /**
     * 投稿を作成
     */
    fun createPost(userId: String): Post {
        val activeItems = getActiveItems()
        return Post(userId = userId, items = activeItems)
    }
    
    /**
     * リセット
     */
    fun reset() {
        _postItems.value = emptyList()
        _currentItemType.value = PostItemType.SCREENSHOT
        _isCreatingPost.value = false
    }
    
    /**
     * 投稿可能かチェック（最低1つのスクリーンショットが必要）
     */
    val canSubmit: Boolean
        get() {
            val activeItems = getActiveItems()
            return activeItems.any { it.type == PostItemType.SCREENSHOT }
        }
}

File: web/src/viewmodels/PostCreationViewModel.ts
/**
 * PostCreationViewModel.ts
 * DailySNSWithYourSS Web版
 * 
 * 投稿作成のビジネスロジック
 */

import { Post, PostItem, PostItemType, createPost, createPostItem, getActiveItems } from '../models/Post';

export class PostCreationViewModel {
  private _postItems: PostItem[] = [];
  private _isCreatingPost: boolean = false;
  private _currentItemType: PostItemType = PostItemType.Screenshot;
  private listeners: Set<() => void> = new Set();

  get postItems(): PostItem[] {
    return this._postItems;
  }

  get isCreatingPost(): boolean {
    return this._isCreatingPost;
  }

  get currentItemType(): PostItemType {
    return this._currentItemType;
  }

  /**
   * 変更を監視
   */
  subscribe(listener: () => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private notify(): void {
    this.listeners.forEach(listener => listener());
  }

  /**
   * スクリーンショットを追加
   */
  addScreenshot(imageURL: string): void {
    const item = createPostItem(PostItemType.Screenshot, imageURL);
    this._postItems.push(item);
    this._currentItemType = PostItemType.Comment;
    this.notify();
  }

  /**
   * コメントを追加
   */
  addComment(text: string): void {
    if (!text.trim()) return;
    const item = createPostItem(PostItemType.Comment, text);
    this._postItems.push(item);
    this._currentItemType = PostItemType.Screenshot;
    this.notify();
  }

  /**
   * アイテムを削除（インデックス指定）
   */
  removeItemAt(index: number): void {
    if (index >= 0 && index < this._postItems.length) {
      this._postItems.splice(index, 1);
      this.notify();
    }
  }

  /**
   * アイテムを削除（ID指定）
   */
  removeItem(id: string): void {
    const index = this._postItems.findIndex(item => item.id === id);
    if (index !== -1) {
      this._postItems.splice(index, 1);
      this.notify();
    }
  }

  /**
   * アクティブなアイテムを取得
   */
  private getActiveItems(): PostItem[] {
    return this._postItems.filter(item => !item.isDeleted);
  }

  /**
   * 投稿を作成
   */
  createPost(userId: string): Post {
    const activeItems = this.getActiveItems();
    return createPost(userId, activeItems);
  }

  /**
   * リセット
   */
  reset(): void {
    this._postItems = [];
    this._currentItemType = PostItemType.Screenshot;
    this._isCreatingPost = false;
    this.notify();
  }

  /**
   * 投稿可能かチェック（最低1つのスクリーンショットが必要）
   */
  get canSubmit(): boolean {
    const activeItems = this.getActiveItems();
    return activeItems.some(item => item.type === PostItemType.Screenshot);
  }
}
